luda-editor/client/src/pages/graphic_asset_manage_page/mod.rs

```diff
@@ -0,0 +1,213 @@
+        cg_list::CgList,
+        cg_viewer::CgViewer,
+        image_list::ImageList,
+        image_viewer::ImageViewer,
+        side_bar::SideBar,
+        top_bar::TopBar,
+        upload_asset::{add_new_image, upload_file},
+    },
+};
+use futures::join;
+use namui::prelude::*;
+use namui_prebuilt::{simple_rect, table::hooks::*};
+use rpc::data::{CgFile, ImageWithLabels};
+use std::ops::Deref;
+
+static TAB_ATOM: Atom<Tab> = Atom::uninitialized_new();
+static IMAGES_ATOM: Atom<Vec<ImageWithLabels>> = Atom::uninitialized_new();
+static CG_FILES_ATOM: Atom<Vec<CgFile>> = Atom::uninitialized_new();
+static SELECTED_ASSET_ATOM: Atom<Option<SelectedAsset>> = Atom::uninitialized_new();
+
+#[namui::component]
+pub struct GraphicAssetManagePage {
+    pub wh: Wh<Px>,
+    pub project_id: Uuid,
+}
+
+impl Component for GraphicAssetManagePage {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        const TOP_BAR_HEIGHT: Px = px(48.0);
+        const SIDE_BAR_WIDTH: Px = px(192.0);
+
+        let Self { wh, project_id } = self;
+
+        let project_id = ctx.track_eq(&project_id);
+        let (tab, _set_tab) = ctx.atom_init(&TAB_ATOM, || Tab::Image);
+        let (_, _) = ctx.atom_init(&IMAGES_ATOM, Vec::new);
+        let (_, _) = ctx.atom_init(&CG_FILES_ATOM, Vec::new);
+        let (selected_asset, _) = ctx.atom_init(&SELECTED_ASSET_ATOM, || None);
+
+        ctx.effect("Fetch graphic assets every project_id changes", || {
+            start_fetch_graphic_assets(*project_id);
+        });
+
+        let on_viewer_close = || {
+            SELECTED_ASSET_ATOM.set(None);
+        };
+        ctx.compose(|ctx| {
+            let Some(selected_asset) = selected_asset.deref() else {
+                return;
+            };
+
+            match selected_asset {
+                SelectedAsset::Image(image) => {
+                    ctx.add_with_key(
+                        "image_viewer",
+                        ImageViewer {
+                            wh,
+                            image,
+                            project_id: *project_id,
+                            on_close: &on_viewer_close,
+                        },
+                    );
+                }
+                SelectedAsset::Cg(cg_file) => {
+                    ctx.add_with_key(
+                        "cg_viewer",
+                        CgViewer {
+                            wh,
+                            cg_file,
+                            project_id: *project_id,
+                            on_close: &on_viewer_close,
+                        },
+                    );
+                }
+            }
+        });
+
+        ctx.compose(|ctx| {
+            vertical([
+                fixed(TOP_BAR_HEIGHT, |wh, ctx| {
+                    ctx.add(TopBar {
+                        wh: Wh::new(wh.width, TOP_BAR_HEIGHT),
+                        project_id: *project_id,
+                    });
+                }),
```

Comment:
This code gave me some inspiration that table fns like `fn fixed` can get trait which is impl-ed by FnOnce(wh, ctx) and FnOnce(wh).